docs/src/design/expressions.md

```diff
@@ -0,0 +1,130 @@
+    MapLibre offers expressions that can interpolate, and paint. An
+    expression can be applied to a feature constantly, zoom dependent,
+    property dependent, or in zoom and property dependent manner.
+
+Expressions use a JSON-like syntax to define. For brevity, let's look at
+an example expression below:
+
+```
+'^': [
+NumberType,
+NumberType, NumberType],
+(ctx, [b, e]) => Math.pow(b.evaluate(ctx), e.evaluate(ctx))
+]
+```
+
+This defines an expression named `^` that returns a number
+expression, and takes two number expressions as input. The
+implementation follows right after. Another part to notice here is the
+implementation evaluates both inputs because they are expressions too.
+Expressions can be nested.
+
+**Although it looks like JavaScript, for MapLibre Native GL**, **the
+parser for any expression is written in MapLibre GL Native Core**. Each
+platform such as iOS, Android has their own Expression class which
+provides builders to build an expression and add it to a layer. When an
+expression is added to a layer, the rendering part of the code picks it
+up in the MapLibre Native GL Core.
+
+Also, inside MapLibre GL Core, this definition mechanism allows
+extending expressions library with custom expressions if desired.
+
+## Expression Types
+
+In the example expression, we saw how one expression is defined. The
+example also shows that expressions have types. Expression language can
+accept input and output types of null, number, string, boolean, color,
+object, value, array, error, collator, and formatted.
+
+Expressions are of different *kinds* such as assertion, coalesce,
+interpolate, distance etc. Each *kind* of expression performs a single
+responsibility.
+
+*Assertion expressions* assert the returning type from one expression is
+asserted before putting into another. For example, a filter expression
+of `["get", "feature_property"]`, returns a generic *value* type. To
+use it on another expression that accepts a string type, an assertion
+such as `["string", ["get", "feature_property"]]` is necessary.
+Assertion throws an evaluation-time error if the types don't match
+during evaluation.
+
+*Coercion expressions* convert a return type to another type. It also
+allows to define a fallback if the conversion fails. A good example of
+this is *to-number* expression. For example, `["to-number", ["get",
+"feature_property"], 0]` means that we are trying to cover the
+feature-property to a number. If it fails, we will use 0 as a fallback.
+
+*Camera expressions* are expressions that allow style property
+manipulation based on zoom, pitch, and distance from center.
+
+Expressions are used *paint* property of a Mapbox Vector Style starting
+with a *layer* selector. Expressions are evaluated by following an
+expression chain constructed from the root expression following the
+input expressions recursively.
+
+## Implementation
+
+Implementation wise, expressions are divided into builders and parsers.
```

Comment:
It would be really useful here to include links to the classes in iOS/Android/core/JS. I know the files may move around, but it'd still be incredibly helpful for someone starting out who wants to extend expressions to know what the class names are/where they may be found.

Reply:
Oh I wanted to add the classes here honestly. But my knowledge is very limited because I am new to this field. MapLibre Native asks for Graphics, C++, iOS, Android, and Native Interop knowledge all to be have full convergeance of knowledge.

That is why I refrained from adding class names and code pointers. The more I learn, I will update this document with code pointers. I sadly can not act on this feedback now. :)